app/src/main/java/com/dkanada/gramophone/service/QueueManager.java

```diff
@@ -0,0 +1,314 @@
+        shuffledQueue.addAll(songs);
+
+        resetCurrentSong = false;
+        callbacks.onQueueChanged();
+    }
+
+    public void removeSong(int position) {
+        if (getShuffleMode() == SHUFFLE_MODE_NONE) {
+            playingQueue.remove(position);
+            shuffledQueue.remove(position);
+        } else {
+            playingQueue.remove(shuffledQueue.remove(position));
+        }
+
+        int currentPosition = getPosition();
+        if (position != currentPosition) {
+            resetCurrentSong = false;
+        }
+
+        if (position < currentPosition) {
+            this.position = currentPosition - 1;
+        }
+
+        callbacks.onQueueChanged();
+    }
+
+    public void moveSong(int from, int to) {
+        if (from == to) return;
+
+        final int currentPosition = getPosition();
+        Song songToMove = getPlayingQueue().remove(from);
+        getPlayingQueue().add(to, songToMove);
+
+        if (from > currentPosition && to <= currentPosition) {
+            position = currentPosition + 1;
+        } else if (from < currentPosition && to >= currentPosition) {
+            position = currentPosition - 1;
+        } else if (from == currentPosition) {
+            position = to;
+        }
+
+        resetCurrentSong = false;
+        callbacks.onQueueChanged();
+    }
+
+    public void clearQueue() {
+        playingQueue.clear();
+        shuffledQueue.clear();
+
+        position = 0;
+        callbacks.onQueueChanged();
+    }
+
+    public long getQueueDurationMillis(int position) {
+        long duration = 0;
+        for (int i = position + 1; i < playingQueue.size(); i++) {
+            duration += playingQueue.get(i).duration;
+        }
+
+        return duration;
+    }
+
+    public void cycleRepeatMode() {
+        switch (getRepeatMode()) {
+            case REPEAT_MODE_NONE:
+                setRepeatMode(REPEAT_MODE_ALL);
+                break;
+            case REPEAT_MODE_ALL:
+                setRepeatMode(REPEAT_MODE_THIS);
+                break;
+            default:
+                setRepeatMode(REPEAT_MODE_NONE);
+                break;
+        }
+    }
+
+    public void restoreQueue() {
+        position = PreferenceUtil.getInstance(context).getPosition();
+        restoredProgress = PreferenceUtil.getInstance(context).getProgress();
+
+        playingQueue = new ArrayList<>(App.getDatabase().queueSongDao().getQueue(0));
+        shuffledQueue = new ArrayList<>(App.getDatabase().queueSongDao().getQueue(1));
+
+        shuffleMode = PreferenceUtil.getInstance(context).getShuffle();
+        repeatMode = PreferenceUtil.getInstance(context).getRepeat();
+
+        callbacks.onQueueChanged();
+        callbacks.onRepeatModeChanged();
+        callbacks.onShuffleModeChanged();
+    }
+
+    public void saveQueue() {
+        PreferenceUtil.getInstance(context).setPosition(position);
+        // copy queues by value to avoid concurrent modification exceptions from database
```

Comment:
I prefer newlines above comments for legibility, although this might have been copied directly from the old code.